src/extractor/location_dependent_data.cpp

```diff
@@ -0,0 +1,284 @@
+
+        const auto y_min = envelop.min_corner().y();
+        const auto y_max = envelop.max_corner().y();
+        const auto dy = (y_max - y_min) / num_bands;
+
+        for (const auto &segment : segments)
+        {
+            using coord_t = boost::geometry::traits::coordinate_type<point_t>::type;
+            const std::pair<coord_t, coord_t> mm =
+                std::minmax(segment.first.y(), segment.second.y());
+            const auto band_min = std::min<coord_t>(num_bands - 1, (mm.first - y_min) / dy);
+            const auto band_max = std::min<coord_t>(num_bands, ((mm.second - y_min) / dy) + 1);
+
+            for (auto band = band_min; band < band_max; ++band)
+            {
+                bands[band].push_back(segment);
+            }
+        }
+
+        polygons.emplace_back(std::make_pair(bands, properties_index));
+    };
+
+    for (rapidjson::SizeType ifeature = 0; ifeature < features_array.Size(); ifeature++)
+    {
+        util::validateFeature(features_array[ifeature]);
+        const auto &feature = features_array[ifeature].GetObject();
+        const auto &geometry = feature["geometry"].GetObject();
+        BOOST_ASSERT(geometry.HasMember("type"));
+
+        // Case-sensitive check of type https://tools.ietf.org/html/rfc7946#section-1.4
+        if (std::strcmp(geometry["type"].GetString(), "Polygon") == 0)
+        {
+            // Collect feature properties and store in polygons vector
+            auto properties_index = collect_properties(feature["properties"].GetObject());
+            const auto &coordinates = geometry["coordinates"].GetArray();
+            index_polygon(coordinates, properties_index);
+        }
+        else if (std::strcmp(geometry["type"].GetString(), "MultiPolygon") == 0)
+        {
+            auto properties_index = collect_properties(feature["properties"].GetObject());
+            const auto &polygons = geometry["coordinates"].GetArray();
+            for (rapidjson::SizeType ipolygon = 0; ipolygon < polygons.Size(); ++ipolygon)
+            {
+                index_polygon(polygons[ipolygon].GetArray(), properties_index);
+            }
+        }
+    }
+
+    // Create R-tree for bounding boxes of collected polygons
+    rtree = rtree_t(bounding_boxes);
+    util::Log() << "Parsed " << properties.size() << " location-dependent features with "
+                << polygons.size() << " GeoJSON polygons";
+}
+
+LocationDependentData::properties_t LocationDependentData::operator()(const point_t &point) const
+{
+    properties_t result;
+
+    auto merger = [this, &result](const rtree_t::value_type &rtree_entry) {
+        const auto &polygon_properties = properties[polygons[rtree_entry.second].second];
+        result.insert(polygon_properties.begin(), polygon_properties.end());
+    };
+
+    // Search the R-tree and collect a Lua table of tags that correspond to the location
+    rtree.query(boost::geometry::index::intersects(point) &&
```

Comment:
Instead of saving each polygon in one bounding box you could try a quad tree approximation:
- Divide the bounding box into 4 equal parts
- For every new box check if its fully container, fully outside or intersects the boundary
- If a box intersects the boundary: Continue dividing that box.
- Stop when all boxes are either fully in or fully out, or you reached a minimum box size (like 10x10m)

Now your query can terminate immediately when it hits a fully in/fully out bounding box. You only need to do the expensive point-in-polygon check on the points that hit the boundary boxes.